web/packages/teleport/src/Discover/Server/TestConnection/useTestConnection.ts

```diff
@@ -1,69 +0,0 @@
-/**
- * Copyright 2022 Gravitational, Inc.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-import cfg from 'teleport/config';
-import { openNewTab } from 'teleport/lib/util';
-import { useConnectionDiagnostic } from 'teleport/Discover/Shared';
-
-import { NodeMeta } from '../../useDiscover';
-
-import type { AgentStepProps } from '../../types';
-import type { MfaAuthnResponse } from 'teleport/services/mfa';
-
-export function useTestConnection(props: AgentStepProps) {
-  const { runConnectionDiagnostic, ...connectionDiagnostic } =
-    useConnectionDiagnostic();
```

Comment:
Regarding "reverse hook drilling": half of what `useTestConnection` does is calling `useConnectionDiagnostic` to pass its return values up.

Instead, `TestConnection` can call both `useConnectionDiagnostic` and `useTestConnection`. If `useTestConnection` needs some stuff that `useConnectionDiagnostic` returns, `TestConnection` can pass it as an argument to `useTestConnection`.

This makes it easier to see what hooks actually depend on and makes it easier to reuse those hooks in different places.

In this case, I removed `useTestConnection` altogether and move its contents to `TestConnection` directly. The only two other things it was doing were:

* Creating `startSshSession` and `testConnection`: these can be shared as functions once I get around to implementing `TestConnection` for Connect My Computer.
* Preparing logins: this was inlined and logins will be fetched through other means in the Connect My Computer connection test.

Reply:
The same pattern exists in `web/packages/teleport/src/Discover/Kubernetes/TestConnection/useTestConnection.ts` and `web/packages/teleport/src/Discover/Database/TestConnection/useTestConnection.t`, do you plan on refactoring those as well?

Reply:
I don't, I wanted to refactor the SSH component so that I can reuse some parts for Connect My Computer connection test more easily.

But I'll add a deprecation comment to those hooks with a link to this PR.